### ISSUE TITLE
Qt: View: Holding a key down retriggers key down/up actions indefinitely, which can't be identified as 'repeats'

### ISSUE THREAD

**Issue by SimonDeplat** [supercollider/supercollider#6466]
## Environment
SC 3.14.0-dev | Ubuntu 22.04LTS | Qt5
Likely concerns every environment.

## Steps to reproduce

```
(
UserView()
.keyDownAction_({ "pressed".postln; })
.keyUpAction_({ "released".postln; })
.front;
)
```
or
```
(
View.globalKeyDownAction_({ "pressed".postln; });
View.globalKeyUpAction_({ "released".postln; });
UserView().front;
)
```
Then hold any key down.

## Expected vs. actual behavior
Holding key down will retrigger key up and down actions indefinitely after a short delay. This behavior is kind of expected, however, this doesn't allow to discriminate between holding key down, and actually releasing it down and pressing it again. Which prevents using the computer keyboard as a 'synthesizer keyboard', for example, when using Views.

Two expected behavior are envisageable: 

- Disabling keyDown and keyUp events emission when key is held.
- Adding a boolean argument to those methods that identifies if the event was triggered by a 'held key' input.

## Context
This issue was discussed on the forum in 2018 by user _simdax_. He first proposed [a sclang side hack](https://scsynth.org/t/cancel-key-repeat-in-gui/319), then [a C++ side quick fix](https://scsynth.org/t/qt-add-repeat-touch-option/321). Both solutions seem unreliable at first glance.

**Comment by capital-G** [comment]
I think a feasible solution would be to add an additional state?

```supercollider
var pressed = false;
UserView()
.keyDownAction_({
	if(pressed.not, {
		pressed = true;
		"pressed".postln;
	});
})
.keyUpAction_({
	pressed = false;
	"released".postln;
})
.front;
```

Would this solve your problem?

**Comment by SimonDeplat** [comment]
This behaves exactly the same as my first example. Holding a key down still posts "released pressed released pressed released pressed..." as long as you hold the key down.

I'll express the problem in an other way : `keyUpAction` can be evaluated without a key being released.

To test this, clear post window, evaluate the code you provided using the mouse (Language > Evaluate File), then press any key and hold it down. Your post window will output "released pressed released pressed released pressed...", with "released" meaning the `keyUpAction` has been evaluated. Still, you never released a key since the program has been started.

**Comment by SimonDeplat** [comment]
Yes I'm using Gnome.

Good to know this is an OS related issue. I believed this was only related to the Qt implementation. I can't test it on Windows tho.

I found documented similar issues, on other C++ projects. I'll have a look at #3972 to see if I can resume what have been done already and propose a fix, however, I'm uncomfortable with C++, so I might need some help concerning the potential multi-OS aspect of the issue.

**Comment by SimonDeplat** [comment]
I can confirm this is related to X11, reference post about this general issue [can be found here](https://stackoverflow.com/questions/2100654/ignore-auto-repeat-in-x11-applications). Still investigating.

**Comment by capital-G** [comment]
> First, of all, can you test this on macOS:

It works as one would expect - pressing `e` for a long time will result in `eeeeeeeeeeeee` and the up and down arrow can be used to adjust the slider without a need to repeatedly hit the key.
As your proposed code, which makes the modification of slider by holding down a key inactive IIUC, is a breaking change I think this would not be the way to go.

The PR could be a way, but I think it should be expanded in such a way that it creates consistent behavior across all platforms? This could be made possible by using the passed `repeat` value within the methods `keyUp` and `keyDown` to perform a lookup for the Linux [Platform](https://dev.docs.supercollider.online/Classes/Platform.html) through a [Set](https://dev.docs.supercollider.online/Classes/Set.html). This makes it possible to filter out any excessive "release/press" events. Still a *hack*, but I think it could be a way to fix this as I don't see any mentions or a fix for OS discrepancies in https://doc.qt.io/qt-5/qkeyevent.html - but there is also https://bugreports.qt.io/browse/QTBUG-57335 😕 

I think it still would be valid to forward the new `repeat` argument to the `keyDownEvent` / `keyUpEvent` implementation (though does the later make sense?).

**Comment by capital-G** [comment]
Just tested this on fedora with wayland where it also does the unintended behavior, so this is not a x11 exclusive issue.

**Comment by SimonDeplat** [comment]
> I think it should be expanded in such a way that it creates consistent behavior across all platforms?

I totally agree. However, this might be strictly impossible because different OS respond differently to key repeats.

I have difficulties understanding the fix you are proposing. Are you proposing that `repeat` is passed to `sclang`, but that end user doesn't see it when using `mouseDownAction` ?

Also, I think you are proposing to discard repeats when OS==Linux. This would fix `View` retriggering the Up/Down loop, but this would also prevent Linux user to increment `Sliders` while holding a key down, unless we manually specify for each class how they should respond to repeats?
